### ISSUE TITLE
Nullable field is incorrectly recognized as non-nullable

### ISSUE THREAD

**Issue by tomwyr** [rrousselGit/freezed#409]
**Describe the bug**
Code generation fails to generate constructor that contains parameter that is at the same time:
- derived type of another freezed class
- nullable
- has no default value
- is not required

`build_runner` output error:
```
[SEVERE] freezed:freezed on lib/a.dart:

The parameter `b` of `A` is non-nullable but is neither required nor marked with @Default
package:fix_freezed/a.dart:9:12
  ╷
9 │     BImpl? b,
  │            ^
  ╵
```

If `b` is declared in any of the following ways, error disappears and classes are successfully generated:
```Dart
B? b
required BImpl? b
@Default(null) BImpl? b
```

**To Reproduce**

`b.dart`
```Dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'b.freezed.dart';

@freezed
class B with _$B {
  factory B() = BImpl;
}
```
`a.dart`
```Dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'b.dart';

part 'a.freezed.dart';

@freezed
class A with _$A {
  factory A({
    BImpl? b,
  }) = _A;
}
```

**Expected behavior**
Code generation creates class `A` that contains `BImpl?` type field and makes it optional in `A` constructor.

**Comment by tomwyr** [comment]
Also this looks similar to #399, however the use case seems to be slightly different.

**Comment by SinaArdehali** [comment]
I have the same problem, and i cannot understand why. I created a blank project using he following code that i found on the official https://pub.dev/packages/freezed#how-to-use :

```
@freezed
class Union with _$Union {
  const factory Union(int value) = Data;
  const factory Union.loading() = Loading;
  const factory Union.error([String message]) = ErrorDetails;
}
```

and my pubspec.yml:

```
dependencies:
  flutter:
    sdk: flutter
  freezed_annotation: ^0.14.1

dev_dependencies:
  flutter_test:
    sdk: flutter
  build_runner: ^1.12.2
  freezed: ^0.14.1+1
```

I get the following error:

```

The parameter `message` of `Union.error` is non-nullable but is neither required nor marked with @Default
package:essai_freezed/main.dart:17:37
   ╷
17 │   const factory Union.error([String message]) = ErrorDetails;
   │                                     ^^^^^^^
   ╵

```

**Comment by iamarnas** [comment]
I working with a json to dart generator and tried to integrate support for Dart null-safety but my result was the same errors. Now I am sure that this error does not come from my side and now I am forced to inform users that Freezed has this error 😔

**Comment by rrousselGit** [comment]
Make sure that your file has null-safety enabled and that your dependencies are up to date.

**Comment by iamarnas** [comment]
@rrousselGit Hi,

Have you time to explain? I use latest dependencies and Dart 2.12 with Flutter 2.0 it was not issues with it. I configured my generator to work with `Freezed` null safety support. I got error because Freezed not friendly with type `dynamic` after I changed all `dynamic` type to `Object` `Freezed` generated without issues. Whats wrong with type `dynamic`? Google.dev still use `dynamic` for `json`, example `json_serializable` developers. If `Freezed` not support `dynamic` type it should be written in the documentation, yes? You can see in the pictures `Freezed` not accept `dynamic` type with indicate that a variable might have the value `null` or without it always provide same error.

![dynamic_error](https://user-images.githubusercontent.com/42679871/114675983-20790280-9d09-11eb-9749-b1840c02b058.png)

**Comment by Salahlemangar** [comment]
`[SEVERE] freezed:freezed on lib/src/domain/models/restaurant.dart:

The parameter `id` of `Restaurant` is non-nullable but is neither required nor marked with @Default
package:food_delivery_ui/src/domain/models/restaurant.dart:13:9
   ╷
13 │     int id,
   │         ^^
   ╵`

**Comment by iamarnas** [comment]
@Salahlemangar Hi.
You need indicate that a variable might have the value `null`
```dart
int? id;
```

**Comment by Salahlemangar** [comment]
Hi @iamarnas ,

- I tried to do so but when I do I got syntax maybe because null safety is disabled but also when I try to enable it using this command **dart migrate --apply-changes** I got errors.

Any suggestions please ?

**Comment by iamarnas** [comment]
> Hi @iamarnas ,
> 
> - I tried to do so but when I do I got syntax error maybe because null safety is disabled but also when I try to enable it using this command **dart migrate --apply-changes** I got errors.
> 
> Any suggestions please ?

Probably something wrong with your dependencies. I use Flutter 2.0 and Dart 2.12 with latest dependencies. No problem for me. Just `Freezed` with null-safety not accept primitive `dynamic` types.

**Comment by GearGit** [comment]
> **Describe the bug**
> Code generation fails to generate constructor that contains parameter that is at the same time:
> 
> * derived type of another freezed class
> * nullable
> * has no default value
> * is not required
> 
> `build_runner` output error:
> 
> ```
> [SEVERE] freezed:freezed on lib/a.dart:
> 
> The parameter `b` of `A` is non-nullable but is neither required nor marked with @Default
> package:fix_freezed/a.dart:9:12
>   ╷
> 9 │     BImpl? b,
>   │            ^
>   ╵
> ```
> 
> If `b` is declared in any of the following ways, error disappears and classes are successfully generated:
> 
> ```dart
> B? b
> required BImpl? b
> @Default(null) BImpl? b
> ```
> 
> **To Reproduce**
> 
> `b.dart`
> 
> ```dart
> import 'package:freezed_annotation/freezed_annotation.dart';
> 
> part 'b.freezed.dart';
> 
> @freezed
> class B with _$B {
>   factory B() = BImpl;
> }
> ```
> 
> `a.dart`
> 
> ```dart
> import 'package:freezed_annotation/freezed_annotation.dart';
> 
> import 'b.dart';
> 
> part 'a.freezed.dart';
> 
> @freezed
> class A with _$A {
>   factory A({
>     BImpl? b,
>   }) = _A;
> }
> ```
> 
> **Expected behavior**
> Code generation creates class `A` that contains `BImpl?` type field and makes it optional in `A` constructor.

Hey,

Even I am getting the same issue.

```
part 'expense.freezed.dart';

@freezed
class Expense with _$Expense {
  const factory Expense({
    ExpenseRemark remark,
    required UniqueId id,
    required ExpenseTitle title,
    required ExpenseAmount amount,
  }) = _Expense;
}
```
The error I got:

```
The parameter `remark` of `Expense.` is non-nullable but is neither required nor marked with @Default
   ╷
10 │     ExpenseRemark remark,
   │                   ^^^^^^
   ╵
```

Here I want the remark part to be nullable/optional, but as per `freezed` it has to be non-nullable. I can somehow make the remark nullable by allowing the user to type nothing and by default passing an empty string, but that does not seem to be the clean way to do so. If we can allow a nullable field, that would be awesome. 

My dependencies:

```
environment:
  sdk: ">=2.12.0 <3.0.0"

dependencies:
  flutter:
    sdk: flutter
  freezed_annotation: ^0.14.2

dev_dependencies:
  freezed: ^0.14.2
```
Any ideas?

**Comment by liri2006** [comment]
@GearGit you have to mark your field as nullable or add a required keyword, that is what the error is saying.
This should work:

```
part 'expense.freezed.dart';

@freezed
class Expense with _$Expense {
  const factory Expense({
    ExpenseRemark? remark,
    required UniqueId id,
    required ExpenseTitle title,
    required ExpenseAmount amount,
  }) = _Expense;
}
```

**Comment by vadrian89** [comment]
Hello, I am having the same issue here with dynamic type:
This is my constructor:
```
const factory UserDocsModel({
    String? id,
    int? docTypeId,
    @JsonKey(ignore: true) String? name,
    @JsonKey(ignore: true) String? fileName,
    String? url,
    @JsonKey(ignore: true) String? urlToDelete,
    @JsonKey(fromJson: dateTimeFromJson, toJson: dateTimetoJson) DateTime? dateAdded,
    @JsonKey(fromJson: dateTimeFromJson, toJson: dateTimetoJson) DateTime? dateModified,
    @JsonKey(ignore: true) dynamic? document,
    String? ownerId,
    @JsonKey(ignore: true) UserDocType? docType,
  }) = _UserDocsModel;
```

This is the error:

> The parameter `document` of `UserDocsModel` is non-nullable but is neither required nor marked with @Default

Flutter (Channel stable, 2.2.2, on Linux, locale ro_RO.UTF-8)
In pubspec I have:
environment:
```
  sdk: '>=2.12.0 <3.0.0'

dev_dependencies:
  flutter_test:
    sdk: flutter
  build_runner: ^2.0.4
  flutter_launcher_icons: ^0.9.0
  freezed: ^0.14.2
  json_serializable: ^4.1.3
```
Everything is update to date as I am trying to migrate the project (first one) to null safe.

**Comment by iamarnas** [comment]
@vadrian89 🖐️
 You can correct this option by using an `Object`.

```dart
//..... 
@JsonKey(ignore: true) Object? document,
//..... 
```

**Comment by vadrian89** [comment]
> @vadrian89 
> You can correct this option by using an `Object`.
> 
> ```dart
> //..... 
> @JsonKey(ignore: true) Object? document,
> //..... 
> ```

I saw the workaround and applied it, thank you, but still I think it's a bug and wanted to report.

**Comment by iamarnas** [comment]
@vadrian89 
> > You can correct this option by using an `Object`.
> > 
> > ```dart
> > //..... 
> > @JsonKey(ignore: true) Object? document,
> > //..... 
> > ```
> 
> I saw the workaround and applied it, thank you, but still I think it's a bug and wanted to report.

I think Remi follow new Dart recommendations by preferring `Object` instead `dynamic`. But I personally think `Freezed` should support `dynamic` becouse sometimes we must use it.

**Comment by peterhijma** [comment]
I've got the same problem, similar to the topic starter. Something defined with a question mark still results in the non-nullable error.

**Comment by rrousselGit** [comment]
This was fixed in version 0.14.3 so I will close this
